### ISSUE TITLE
Feature request, Tau Prolog might consider setup_call_cleanup/3

### ISSUE THREAD

**Issue by jburse** [tau-prolog/tau-prolog#71]
I just checked, there is no setup_call_cleanup/3:

![image](https://user-images.githubusercontent.com/5411487/51113049-0df98100-1801-11e9-98ed-97c9819a6d0c.png)

What is setup_call_cleanup/3:

> ISO/IEC JTC1 SC22 WG17 post-N215
> https://www.complang.tuwien.ac.at/ulrich/iso-prolog/cleanup
>

**Comment by jburse** [comment]
I don't know whether Tau Prolog has already a use-case / business-case for it. 
The simplest use case is information integration, like for example accessing
a remote database. The programming pattern:

```
     setup_call_cleanup(
           open_db(...),
           query_db(...),
           close_db(...))

```
The way setup_call_cleanup/3 is defined, the above works also when query_db(...)
is non-deterministic, and close_db(...) will only be called when the last solution
was delivered or when an even happens inside the query or the continuation.

In case a Prolog system decides to provide this predicate, it can be helpful
to work early on towards such a predicate, because it might change the whole
infrastructure of a Prolog interpreter. If I remember well O-Prolog has it, but

there is still some testing going on for O-Prolog. Prolog systems such as SWI, 
ECLiPSe (library(prolog_extras)) etc.. all have it. On the otherhand GNU doesn't 
have it. SICStus shows me only call_cleanup/2. 

So it might or might not be a disposable feature.

**Comment by jariazavalverde** [comment]
Done! It will be available in the next release, v0.3.0. Added the following built-in predicates:

- call_cleanup/2
- setup_call_cleanup/3

Thanks for the info.
